keps/sig-network/2079-network-policy-port-range/README.md

```diff
@@ -0,0 +1,339 @@
+the rule should be created specifying each port separately, as:
+
+```
+spec:
+  egress:
+  - ports:
+    - protocol: TCP
+      port: 32000
+    - protocol: TCP
+      port: 32001
+    - protocol: TCP
+      port: 32002
+    - protocol: TCP
+      port: 32003
+[...]
+    - protocol: TCP
+      port: 32768
+```
+
+So for the user:
+* To allow a range of ports, each of them must be declared as an item from 
+``ports`` array
+* To make an exception needs a declaration of all ports but the exception
+
+Adding a new ``PortRange`` field inside the ``ports`` will allow a simpler 
+creation of NetworkPolicy to the user.
+
+### Goals
+
+Add Range field to Ports in NetworkPolicy
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+In NetworkPolicy specification, inside ``NetworkPolicyPort`` object struct 
+specify a new ``Range`` field composed of the minimum and maximum ports 
+inside the range.
+
+If both ``Port`` and ``Range`` are specified, they are cumulative and no further
+validation might occur. It's up to the CNI to summarize both of the fields in a 
+single rule.
+
+### User Stories (Optional)
+
+#### Story 1
+
+I have an application that communicates with NodePorts of a different cluster 
+and I want to allow the egress of the traffic only the NodePort range 
+(eg. 30000-32767) as I don't know which port is going to be allocated on the 
+other side, but don't want to create a rule for each of them.
+
+### Notes/Constraints/Caveats
+
+*  The technology used by the CNI provider might not support port range in a 
+trivial way as described in [#drawbacks]
+
+### Risks and Mitigations
+
+CNIs will need to support the new field in their controllers. For this case 
+we'll try to make broader communication with the main CNIs so they can be aware
+of the new field.
+
+## Design Details
+
+API changes to NetworkPolicy:
+* Add a new struct called ``NetworkPolicyPortRange`` as the following:
+```
+// NetworkPolicyPortRange describes the range of ports to be used in a 
+// NetworkPolicyPort struct
+type NetworkPolicyPortRange struct {
+    // From defines the start of the port range
+    From         uint16
+    
+    // To defines the end of the port range, being the end included within the
+    // range
+    To           uint16
+
+    // Except defines all the exceptions in the port range
+    Except:      []uint16
+}
+```
+* Add a new field ``spec.ingress|egress.ports.range`` that points to the
+new struct:
+```
+// NetworkPolicyPort describes a port or a range of ports to allow traffic on
+type NetworkPolicyPort struct {
+  // The protocol (TCP, UDP, or SCTP) which traffic must match. If not specified, this
+  // field defaults to TCP.
+  // +optional
+  Protocol *api.Protocol
+  
+  // The port on the given protocol. This can either be a numerical or named 
+  // port on a pod. If this field is not provided but a Range is 
+  // provided, this field is ignored. Otherwise this matches all port names and
+  // numbers.
+  // +optional
+  Port *intstr.IntOrString
```

Comment:
I'm still on the fence about this, but I feel that `Port` and `Range` under the same field feels awkward. Putting range semantics into it's own struct _might_ be cleaner, even if you have to specify the Protocol field twice. This way consumers don't have to reason about precedent between `port` and `range` under the same field, but can just take the union of `ports` and `portRanges`.  

```
ports:
  - protocol: TCP
    port: 80
  - protocol: TCP
    port: 443
portRanges:
  - protocol: TCP
    range:
      from: 30000
      to: 32000
```

Reply:
tbh I have no strong opinion about naming either, so I can move this to outside ports, making it an array and that's fine

Reply:
Let's wait to see what Tim thinks :)

Reply:
Great, thank you for all the feedback Andrew :)

Reply:
I agree `portRanges` is clearer.

Reply:
@andrewsykim @danwinship about @thockin proposal of just adding an EndPort into the existing ``NetworkPolicyPort`` sounds good and this conversation can be resolved?